packages/react/select/src/Select.tsx

```diff
@@ -0,0 +1,1359 @@
+          >
+            <FocusScope
+              asChild
+              // we make sure we're not trapping once it's been closed
+              // (closed !== unmounted when animating out)
+              trapped={context.open}
+              onMountAutoFocus={(event) => {
+                // we prevent open autofocus because we manually focus the selected item
+                event.preventDefault();
+              }}
+              onUnmountAutoFocus={composeEventHandlers(onCloseAutoFocus, (event) => {
+                context.trigger?.focus({ preventScroll: true });
+                event.preventDefault();
+              })}
+            >
+              <DismissableLayer
+                role="listbox"
+                id={context.contentId}
+                data-state={context.open ? 'open' : 'closed'}
+                dir={context.dir}
+                onContextMenu={(event) => event.preventDefault()}
+                {...contentProps}
+                ref={composedRefs}
+                style={{
+                  display: 'flex',
+                  flexDirection: 'column',
+                  // When we get the height of the content, it includes borders. If we were to set
+                  // the height without having `boxSizing: 'border-box'` it would be too big.
+                  boxSizing: 'border-box',
+                  maxHeight: '100%',
+                  outline: 'none',
+                  ...contentProps.style,
+                }}
+                disableOutsidePointerEvents
+                // When focus is trapped, a focusout event may still happen.
+                // We make sure we don't trigger our `onDismiss` in such case.
+                onFocusOutside={(event) => event.preventDefault()}
+                onDismiss={() => context.onOpenChange(false)}
+                onKeyDown={composeEventHandlers(contentProps.onKeyDown, (event) => {
+                  const isModifierKey = event.ctrlKey || event.altKey || event.metaKey;
+                  if (!isModifierKey && event.key.length === 1) handleTypeaheadSearch(event.key);
+
+                  const focusIntent = MAP_KEY_TO_FOCUS_INTENT[event.key];
+                  if (focusIntent !== undefined) {
+                    event.preventDefault();
+                    const items = getItems().filter((item) => !item.disabled);
+                    let candidateNodes = items.map((item) => item.ref.current!);
+
+                    if (['prev', 'last'].includes(focusIntent)) {
+                      candidateNodes = candidateNodes.slice().reverse();
+                    }
+                    if (['prev', 'next'].includes(focusIntent)) {
+                      const currentElement = event.target as SelectItemElement;
+                      const currentIndex = candidateNodes.indexOf(currentElement);
+                      candidateNodes = candidateNodes.slice(currentIndex + 1);
+                    }
```

Comment:
i had commented on @andy-hook's navigation stuff about the focus intent abstractions here, i've been finding it difficult to pick apart. 

we changed andy's navigation to check the keys inline, so something along these lines in your case:

```jsx
if (['ArrowUp', 'ArrowDown', 'Home', 'End'].includes(event.key)) {
  const items = getItems().filter((item) => !item.disabled);
  let candidateNodes = items.map((item) => item.ref.current!);
  if (['ArrowUp', 'End'].includes(event.key)) {
    candidateNodes = candidateNodes.slice().reverse();
  }
  if (['ArrowUp', 'ArrowDown'].includes(event.key)) {
    const currentElement = event.target as SelectItemElement;
    const currentIndex = candidateNodes.indexOf(currentElement);
    candidateNodes = candidateNodes.slice(currentIndex + 1);
  }
  event.preventDefault();
}
```

thoughts?

Reply:
Ah yeah, the abstraction came from `RovingFocusGroup` I think, but probably isn't needed here as there's no special handling for orientation or RTL. It's just a bit trickier inline when there's these 2 cases to handle.

I'll take a look at what you guys did and will update here.

Reply:
Addressed here: #1179